### PR TITLE
 Set attributes only on queue pairs of the correct type 

### PR DESCRIPTION
**Pull request by YtvwlD**
Some of the Queue Pair attributes are only valid for connections and some are only valid for reliable connections.
Setting an attribute that is not valid, might produce `EINVAL` (at least, it did for me. I'm not sure from the man pages whether this always happens).

So, what I've done is add setters for those fields that did not already have one and then, convert those fields to `Option`s and add `panic`s to those setters.

I'm not quite satisfied with panicking and I already thought about replacing the fields (and the type field) with two optional structs named `Connection` and `Reliability` (?), but then I don't know how to set them. Plus, the approach I took is compatible with existing code.

I only modified the fields on `PreparedQueuePair` and `QueuePairBuilder`. This is enough to get a UC QP up and running, for UD, the `Endpoint` struct would probably need to change.

### CODE REVIEW

**Comment by jonhoo** [comment]
Good catch, thanks for reporting back!

I actually wonder if we want to go a slightly different route here: don't panic in the setter, but instead document that these options "do nothing" on inappropriate types, and then keep the `Option` + only set if `Some` behavior you have in here. I think that's the path that's least likely to break existing users.

For the next major version, what I would probably do is parameterize the `PreparedQueuePair` and `QueuePairBuilder` by a generic parameter `QPT`, which can be either `struct Connection` or `struct ReliableConnection`. You could then put all the settings that are only valid for particular types of queue pairs inside those structs, and require that they be configured that way. Doing it like this would ensure that no invalid options are set at compile time, rather than relying on runtime checks and panics or `Result`s.

Which raises the question: would you be interested in picking that effort up? I'm not actually using this crate myself any more, so am arguably not the best custodian for it. I'd be happy to give you commit access to the repo, and eventually publish rights to crates.io!

**Comment by YtvwlD** [comment]
Hm, I think just doing nothing in the setters would be confusing. But I would be fine with `warn!` instead of `panic!`.

I really like the idea of ensuring the correctness at compile time, though I don't think distinguishing between just `Connection` and `ReliableConnection` would be a good way forward. There's quite a bit of overlap there and still, there would be no support for `UD` queue pairs. I was thinking more about having a field `PreparedQueuePair` and `QueuePairBuilder` in the form of a `Option<Connection>`:
If it's not there, it's `UD`. And this struct could have an additional `Option<Reliability>` which would differentiate between `UC` and `RC`.
I think this would be a relatively exact mapping, but I'm not sure how ergonomic these nested structs would be. There would probably need to be builders for each of them.

With regards to the commit rights: I'm just taking my first steps with RDMA and `ibverbs` at the moment (and the documentation of this crate really helped, thank you!), so I'm not sure I would be the right fit (though I'm hoping to get better at this in the next weeks). I have [three very small crates](https://crates.io/users/YtvwlD) at crates.io, so I know the basic stuff, but just that.

So, for the next steps, I'm going to replace the `panic`s I've added with `warn`s. :)

**Comment by jonhoo** [comment]
Ah, so, to be clear, I wasn't proposing separate types for them, I was proposing a single generic type with type _parameters_ for things like reliability. So, for example, `PreparedQueuePair<Connection, Reliable>` or `PreparedQueuePair<Unidirectional, Unreliable>`.

I'd like to avoid taking a dependency on `log` (or a logging crate) just for this. Besides, those logs will still almost certainly be missed by consumers anyway since nothing will happen if they haven't set up logging! So let's leave them out and go with the "silently ignored" (+ documentation!) for now :)

**Comment by YtvwlD** [comment]
> Ah, so, to be clear, I wasn't proposing separate types for them, I was proposing a single generic type with type parameters for things like reliability. So, for example, `PreparedQueuePair<Connection, Reliable>` or `PreparedQueuePair<Unidirectional, Unreliable>`.

Ah yes, that might work. I might have to think a bit about this. :)

> I'd like to avoid taking a dependency on log (or a logging crate) just for this. Besides, those logs will still almost certainly be missed by consumers anyway since nothing will happen if they haven't set up logging! So let's leave them out and go with the "silently ignored" (+ documentation!) for now :)

You're probably right about people not seeing warnings. :(

I'm still not entirely happy with function calls that do nothing, but then again, it's better than erroring with a non-descript `EINVAL`.

**Comment by YtvwlD** [comment]
This is a no-change rebase. Please ping me if I should change something. :)

File: ibverbs/src/lib.rs
Comment (jonhoo) [inline ibverbs/src/lib.rs:653]:
```suggestion
                self.access.expect("always set to Some in new")
```

File: ibverbs/src/lib.rs
Comment (jonhoo) [inline ibverbs/src/lib.rs:809]:
```suggestion
    /// This defaults to 1.
    /// Valid only for RC QPs.
```

File: ibverbs/src/lib.rs
Comment (jonhoo) [inline ibverbs/src/lib.rs:819]:
```suggestion
    /// Defaults to the port's active_mtu.
    /// Valid only for RC and UC QPs.
```

File: ibverbs/src/lib.rs
Comment (jonhoo) [inline ibverbs/src/lib.rs:838]:
```suggestion
    /// Defaults to 0.
    /// Valid only for RC and UC QPs.
```

**Review by jonhoo — APPROVED** [review]
Looks good to me with some minor nits!

I see you left two TODOs — should they block merging?

**Comment by YtvwlD** [comment]
Ah, yes, good catch. :)

I'm personally fine with leaving the TODOs for the moment, because I currently just need UC and RC. :upside_down_face: 
I also don't really know much about UD, so making this work would be more difficult.

**Comment by jonhoo** [comment]
Then let's get this out! Released as 0.8.1 :tada: 
Thanks :D

**Pull request merged**
